### ISSUE TITLE
Packaging request: Daedalus wallet

### ISSUE THREAD

**Issue by jluttine** [NixOS/nixpkgs#32516]
Wallet for Cardano crypto: https://daedaluswallet.io/

Repository: https://github.com/input-output-hk/daedalus

I'll add this here so if anyone starts working on this, they can mention it here in order to avoid multiple people accidentally working on it at the same time. I might do it myself at some point.

**Comment by dtzWill** [comment]
Related issue on their tracker: https://github.com/input-output-hk/daedalus/issues/492

**Comment by dtzWill** [comment]
Initial nix package version is currently a PR: https://github.com/input-output-hk/daedalus/pull/578 !

**Comment by ixmatus** [comment]
You could do this with elf patching, which is something I'm doing right now. I can put up a pull request tonight. Similar to the Signal Desktop app.

Building Daedalus entirely with Nix is probably a non-trivial project, I tried with a first-pass using `nixfromnpm` but it failed on some dependency version bounds that it couldn't find in the public NPM registry.

**Comment by ixmatus** [comment]
I can confirm that after producing the standalone electron node app for Daedalus and patching the elf binary that the wallet app runs.

Something to be aware of though: Daedalus is not the _whole_ wallet. It is only the frontend app and it needs the cardano-sl backend server running. The macapp and the windows app handle this in one bundle.

So we have two hurdles that make this tough to package in Nix right now:

1. No pre-built binaries for the Daedalus front-end application for Linux that we can, say, `fetchurl` or `fetchFromGitHub`; it must be produced by hand
2. Daedalus is _only_ a front-end application and needs the backend which is provided by `cardano-sl` - this does have Nix build instructions and can be packaged but might be better added as a NixOS module with a systemd unit, at least that's what I've done on my own machine

**Comment by ixmatus** [comment]
@jluttine building the `cardano-sl` node from source is definitely possible. In fact that is how I did it on my NixOS laptop.

Building the Daedalus wallet from source is going to be a significant effort because it will require tooling improvements to whatever method is used to build the node project. I've built node projects with `nixfromnpm` successfully but it required a significant amount of hand-editing of the generated code. The generated package set is also not defined as a fixpoint so overriding a package means overriding any package that depends on it and replacing the dependency with the overridden one.

In short: building cardano-sl from source is no problem and packaging the Daedalus app as a binary package will be just fine.

**Comment by jluttine** [comment]
I tried something like this but it didn't quite work, not sure why:
```
git clone https://github.com/input-output-hk/daedalus
cd daedalus
git checkout 0.8.0
nix-shell -p nodePackages.node2nix
node2nix
nix-build -A package
```

Without checking out the 0.8.0 tag, I got some version bound errors. WIth 0.8.0 I got the error messages shown below.

<details>
<summary>Error tail</summary>

```
/nix/store/xfgjfjcq34fwfbfi1idp2nh81kpigch7-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
/nix/store/xfgjfjcq34fwfbfi1idp2nh81kpigch7-patch-shebangs.sh: line 22: warning: command substitution: ignored null byte in input
./daedalus/installers/data/scripts/dockutil: interpreter directive changed from "/usr/bin/python" to "/nix/store/9n2npaz7xy8b4q0a8k28nr3sjzfy625i-python-2.7.14/bin/python"
./daedalus/installers/data/scripts/postinstall: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/ski5znw2b8bj5qnsncrcq6128hkrwgvq-bash-4.4-p12/bin/bash"
npm WARN prefer global marked@0.3.7 should be installed with -g

> react-polymorph@0.5.4 postinstall /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/react-polymorph
> postinstall-build lib --only-as-dependency


> cucumber-expressions@5.0.7 postinstall /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/cucumber-expressions
> node scripts/postinstall.js


> sha3@1.2.0 install /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/sha3
> node-gyp rebuild

make: Entering directory '/nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/sha3/build'
  CXX(target) Release/obj.target/sha3/src/addon.o
  CXX(target) Release/obj.target/sha3/src/displayIntermediateValues.o
  CXX(target) Release/obj.target/sha3/src/KeccakF-1600-reference.o
  CXX(target) Release/obj.target/sha3/src/KeccakNISTInterface.o
  CXX(target) Release/obj.target/sha3/src/KeccakSponge.o
  SOLINK_MODULE(target) Release/obj.target/sha3.node
  COPY Release/sha3.node
make: Leaving directory '/nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/sha3/build'

> scrypt@6.0.3 preinstall /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/scrypt
> node node-scrypt-preinstall.js

Error: Error: Command failed: ./configure
sed: can't read conftest.err: No such file or directory
cat: conftest.er1: No such file or directory
configure: error: in `/nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/scrypt/scrypt/scrypt-1.2.0':
configure: error: C compiler cannot create executables
See `config.log' for more details


> scrypt@6.0.3 install /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/scrypt
> node-gyp rebuild

make: Entering directory '/nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/scrypt/build'
  SOLINK_MODULE(target) Release/obj.target/copied_files.node
  COPY Release/copied_files.node
  CC(target) Release/obj.target/scrypt_wrapper/src/util/memlimit.o
  CC(target) Release/obj.target/scrypt_wrapper/src/scryptwrapper/keyderivation.o
  CC(target) Release/obj.target/scrypt_wrapper/src/scryptwrapper/pickparams.o
  CC(target) Release/obj.target/scrypt_wrapper/src/scryptwrapper/hash.o
  AR(target) Release/obj.target/scrypt_wrapper.a
  COPY Release/scrypt_wrapper.a
  CC(target) Release/obj.target/scrypt_lib/scrypt/scrypt-1.2.0/lib/crypto/crypto_scrypt.o
  CC(target) Release/obj.target/scrypt_lib/scrypt/scrypt-1.2.0/lib/crypto/crypto_scrypt_smix.o
  CC(target) Release/obj.target/scrypt_lib/scrypt/scrypt-1.2.0/libcperciva/util/warnp.o
  CC(target) Release/obj.target/scrypt_lib/scrypt/scrypt-1.2.0/libcperciva/alg/sha256.o
  CC(target) Release/obj.target/scrypt_lib/scrypt/scrypt-1.2.0/libcperciva/util/insecure_memzero.o
  CC(target) Release/obj.target/scrypt_lib/scrypt/scrypt-1.2.0/lib/scryptenc/scryptenc_cpuperf.o
  AR(target) Release/obj.target/scrypt_lib.a
  COPY Release/scrypt_lib.a
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_common.o
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_params_async.o
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_params_sync.o
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_kdf_async.o
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_kdf_sync.o
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_kdf-verify_sync.o
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_kdf-verify_async.o
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_hash_sync.o
  CXX(target) Release/obj.target/scrypt/src/node-boilerplate/scrypt_hash_async.o
  CXX(target) Release/obj.target/scrypt/scrypt_node.o
  SOLINK_MODULE(target) Release/obj.target/scrypt.node
  COPY Release/scrypt.node
make: Leaving directory '/nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/scrypt/build'

> websocket@1.0.24 install /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/websocket
> (node-gyp rebuild 2> builderror.log) || (exit 0)

make: Entering directory '/nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/websocket/build'
  CXX(target) Release/obj.target/bufferutil/src/bufferutil.o
  SOLINK_MODULE(target) Release/obj.target/bufferutil.node
  COPY Release/bufferutil.node
  CXX(target) Release/obj.target/validation/src/validation.o
  SOLINK_MODULE(target) Release/obj.target/validation.node
  COPY Release/validation.node
make: Leaving directory '/nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/websocket/build'

> electron@1.7.9 postinstall /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/electron
> node install.js


> electron-chromedriver@1.7.1 install /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/electron-chromedriver
> node ./download-chromedriver.js

/nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/node_modules/electron-chromedriver/download-chromedriver.js:16
  if (error != null) throw error
                     ^

Error: getaddrinfo ENOTFOUND github.com github.com:443
    at errnoException (dns.js:27:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:78:26)

npm ERR! Linux 4.9.66
npm ERR! argv "/nix/store/47h9pg6gkss3xh1vk0i1sq1nc8idf9xb-nodejs-4.8.5/bin/node" "/nix/store/47h9pg6gkss3xh1vk0i1sq1nc8idf9xb-nodejs-4.8.5/bin/npm" "--registry" "http://www.example.com" "--nodedir=/nix/store/bdy38g97h606sf5dk7h850bq7xmcs5l4-node-sources" "--production" "rebuild"
npm ERR! node v4.8.5
npm ERR! npm  v2.15.11
npm ERR! code ELIFECYCLE
npm ERR! electron-chromedriver@1.7.1 install: `node ./download-chromedriver.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the electron-chromedriver@1.7.1 install script 'node ./download-chromedriver.js'.
npm ERR! This is most likely a problem with the electron-chromedriver package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./download-chromedriver.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs electron-chromedriver
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls electron-chromedriver
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /nix/store/c1yshx3lcwdp8654kvnlg7va1rbaj8d1-node-daedalus-0.8.0/lib/node_modules/daedalus/npm-debug.log
builder for ‘/nix/store/a7s0d3hjb2ig9pf5qpic3y0nmhjfhr6d-node-daedalus-0.8.0.drv’ failed with exit code 1
error: build of ‘/nix/store/a7s0d3hjb2ig9pf5qpic3y0nmhjfhr6d-node-daedalus-0.8.0.drv’ failed
```
</details>

**Comment by ixmatus** [comment]
@jluttine oh, yeah, I don't use `node2nix` for packaging npm projects anymore because its generated Nix code can't be overridden (unlike that from `nixfromnpm`) and there are a few other nodejs isolation issues I ran into (that I cannot remember). Also the version of node `node2nix` is selecting for you there is old, I don't think Daedalus will build with Node 4.x.

**Comment by jluttine** [comment]
I also tried with `yarn2nix`. I modified the `yarn.lock` file a little bit. The derivation expression:
```
{lib, pkgs, ...}:
with (import /home/jluttine/Workspace/yarn2nix { inherit pkgs; });
mkYarnPackage rec {
  name = "${pname}-${version}";
  pname = "daedalus";
  version = "0.8.0";
  src = /home/jluttine/Workspace/daedalus;
  packageJson = /home/jluttine/Workspace/daedalus/package.json;
  yarnLock = /home/jluttine/Workspace/daedalus/yarn.lock;
}
```

Modified yarn.lock file: https://gist.github.com/jluttine/a577e815d17fc2ae71344c820986fb32

However, I get the following error:
```
[1/4] Resolving packages...
error Couldn't find any versions for "aes-js" that matches "2.1.1" in our cache (possible versions are ""). This is usually caused by a missing entry in the lockfile, running Yarn without the --offline flag may help fix this issue.
```

It seems that the packages aren't found for some reason. I also asked on yarn2nix repo about this: https://github.com/moretea/yarn2nix/issues/49

**Comment by jluttine** [comment]
I'm quite lost with this whole Cardano stuff because I haven't used it at all and it seems to consist of a bunch of pieces whose purpose I don't know. Anyway, I guessed that `cardano-sl` is something quite essential. I wrote the following derivation for it, perhaps this is a good place to ask for comments about it:

```
{lib, pkgs, stdenv, config, ...}:

let
  pname = "cardano-sl";
  version = "1.0.3";
  gitrev = "v${version}";
  source = pkgs.srcOnly {
    stdenv = stdenv;
    name = "${pname}-${version}";
    src = pkgs.fetchFromGitHub {
      owner = "input-output-hk";
      repo = pname;
      rev = gitrev;
      sha256 = "1z2yjkm0qwhf588qnxcbz2d5mxvhqdxawwl8dczfnl47rb48jm52";
    };
  };
in (
  (import "${source}/default.nix") {
    inherit pkgs config gitrev;
  }
).cardano-sl
```

This builds `cardano-node-simple` executable. There's also `cardano-node` in `cardano-sl-wallet`, not sure if I should've built that instead.

I could make a pull request about that if you think it's ok and useful. Not sure what else I need to get some kind of working wallet (possibility to at least receive and maybe send Ada).

**Comment by jluttine** [comment]
I accidentally gave an incorrect error tail for `nix-build -A package` after `node2nix`. The real error tail is:
```
> react-polymorph@0.3.1 postinstall /nix/store/d3i3agpd09x5abvpic2l8b68d7wi8mc1-node-dependencies-daedalus-0.8.0/daedalus/node_modules/react-polymorph
> postinstall-build lib --only-as-dependency

sh: postinstall-build: command not found
```

I wasn't able to try `nixfromnpm` because it didn't compile on my NixOS. It raised an error:
```
In-place registering parsers-0.12.7...
Preprocessing test suite 'doctests' for parsers-0.12.7...

tests/doctests.hs:16:8:
    Could not find module ‘Build_doctests’
    Use -v to see a list of the files searched for.
builder for ‘/nix/store/ba1j04lj41qwn19ghdkvgnqhccc9ds65-parsers-0.12.7.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/3g8lc26nh7ryyjw5wcd3w5ap4sms6xc3-ghc-7.10.2-with-packages.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/3g8lc26nh7ryyjw5wcd3w5ap4sms6xc3-ghc-7.10.2-with-packages.drv’ failed
/run/current-system/sw/bin/nix-shell: failed to build all dependencies
```
Maybe some package that nixfromnpm depends on is broken on NixOS at the moment?

**Comment by ixmatus** [comment]
@jluttine they already have a derivation that builds `cardano-sl`. `cardano-sl` is the "backend" of the wallet, it is the node.

The Daedalus wallet is really two components: a node with a rest API and a "front-end" JS app (named Daedalus).

**Comment by ixmatus** [comment]
I used this, to get a node I could run:

```
nix-build --attr connect.mainnetWallet -o connect-to-mainnet release.nix
```

Note though that script has a bug in it that is fixed by my pull request here: https://github.com/input-output-hk/cardano-sl/pull/2083

**Comment by jluttine** [comment]
@ixmatus Yes, that derivation I wrote uses their derivation. If there is simpler or better method to add cardano-sl to nixpkgs using their derivation, I'd be happy to learn.

**Comment by ixmatus** [comment]
Use the `connect.mainnetWallet` example.

**Comment by ixmatus** [comment]
You should be able to then `./connect-to-mainnet` to run the node; you will then need to run the Daedalus electron app separately, which will connect to it. I was able to build a Linux release of the daedalus front-end application and then patchelf it so it would run on my NixOS; @cleverca22 had a simpler approach which was to get the pre-built Javascript from the MacOS application, then run it with a NixOS electron executable.

**Comment by ixmatus** [comment]
@jluttine this will probably be very helpful too: https://github.com/input-output-hk/daedalus/blob/devops-13/linux.nix#L32-L42

**Comment by ixmatus** [comment]
I haven't had any time to help with packaging, when I have some free-time I'll try to get most of what I've done in a more easily consumable form.

**Comment by jluttine** [comment]
@ixmatus Ok. Packaging of that `connect.mainnetWallet` wasn't as simple because `release.nix` fixes nixpkgs version. Gotta look some other time or if someone more familiar with Cardano and nixpkgs can do the packaging, that'd be great too.

**Comment by ixmatus** [comment]
I'm confused, what's wrong with the pinned `nixpkgs`? What error did you get running:

```
nix-build --attr connect.mainnetWallet -o connect-to-mainnet release.nix
```

**Comment by jluttine** [comment]
@ixmatus I've thought that packages in `nixpkgs` shouldn't use a pinned version of `nixpkgs` but use packages from that same `nixpkgs`. Or are there some packages in `nixpkgs` that use dependencies from a pinned version of `nixpkgs`? Could you provide an example? To me, that would sound a bit weird.

I didn't yet try that command you provided, I'd expect it to run successfully. But I'm trying to package Daedalus for nixpkgs, so it's a slightly different matter, as far as I understand.

**Comment by ixmatus** [comment]
@jluttine I thought you were just trying to build the node. You are correct that packaging the cardano-sl node for `nixpkgs` means we should not use the pinned `nixpkgs` from the cardano-sl repository.

`cardano-sl` is just a Haskell project and you can see its Nix build instructions in `cardano-sl/pkgs/default.nix`. Then there's a script for running the node with some pre-defined configurations which is in this Nix file: `./scripts/launch/connect-to-cluster/default.nix`.

What you will want to do if you want to package the node for `nixpkgs` is get the Nix build instructions for the haskell project isolated. Then, the script can be isolated and added to the ecosystem (really, that mainnet script should be a systemd unit definition).

There is some work involved and I can show you a few examples. However, I do not think an import from derivation is the right thing to do here.

**Comment by ixmatus** [comment]
I want to get my own systemd node up so I'm going to package this real quickly and show you what I'm thinking.

**Comment by domenkozar** [comment]
Current efforts are tracked in https://github.com/input-output-hk/daedalus/issues/492

**Comment by jbgi** [comment]
Does the recent official release of daedalus for linux change the way we should integrate it in nixos? What is your recommendation @domenkozar?

**Comment by cleverca22** [comment]
it could potentially be done, but it wont support the auto-update that the main binary supports
also, due to nixpkgs not allowing import-from-derivation, i would have to add nearly 2mb of nix expressions to nixpkgs

**Comment by jluttine** [comment]
FWIW, Daedalus build is now fully "nixified": https://github.com/input-output-hk/daedalus/issues/492

Not sure if that makes adding it to nixpkgs any easier.

I wonder why and how is Daedalus so different from all other crypto software like bitcoin, monero etc, so that its addition to nixpkgs is so much more difficult. Feels weird.

**Comment by jluttine** [comment]
Still would love to see Daedalus in nixpkgs similarly as many other crypto wallets..

**Comment by asbachb** [comment]
Still relevant.

**Comment by Fresheyeball** [comment]
Holy shit what in the actual fuck why is this not done?

**Comment by ejpcmac** [comment]
Hi @Fresheyeball, maybe because the packaging is done by persons like you and me on our free time, so someone has to do it. I’m also very interesting in this, and looking into it is somewhere on my TODO list (as it may be on someone elses’ also), but I do other things because there are also other things on my TODO list that make me more happy or are more urgent to do at the time.

If it seems so strange that it has not been done, whether because it is extremely important for you or you think it is really easy to do, maybe you could open a PR with the packaging? I’m sure you’ll make people happy :)

**Comment by Fresheyeball** [comment]
@ejpcmac https://github.com/NixOS/nixpkgs/pull/148354

I was really just shocked it hasn't been done yet. Happy to help here.

**Comment by jluttine** [comment]
Still relevant to me. (Re: stale status by the bot)
